node/src/accountant/payment_adjuster/preparatory_analyser/mod.rs

```diff
@@ -0,0 +1,665 @@
+        let mut weighted_account_1 = make_weighed_account(123);
+        weighted_account_1
+            .analyzed_account
+            .qualified_as
+            .bare_account
+            .balance_wei = balance_1;
+        let balance_2 = multiple_by_billion(3_456_000);
+        let mut weighted_account_2 = make_weighed_account(456);
+        weighted_account_2
+            .analyzed_account
+            .qualified_as
+            .bare_account
+            .balance_wei = balance_2;
+        let accounts = vec![weighted_account_1, weighted_account_2];
+        let service_fee_totally_required_minor = balance_1 + balance_2;
+        let cw_service_fee_balance_minor = service_fee_totally_required_minor + 1;
+        let error_factory = LateServiceFeeSingleTxErrorFactory::new(&accounts);
+        let logger = Logger::new(test_name);
+        let subject = PreparatoryAnalyzer::new();
+
+        [(0, false), (1, false), (2, true)].iter().for_each(
+            |(subtrahend_from_cw_balance, expected_result)| {
+                let service_fee_balance = cw_service_fee_balance_minor - subtrahend_from_cw_balance;
+                let result = subject
+                    .recheck_if_service_fee_adjustment_is_needed(
+                        &accounts,
+                        service_fee_balance,
+                        error_factory.clone(),
+                        &logger,
+                    )
+                    .unwrap();
+                assert_eq!(result, *expected_result);
+            },
+        );
+        TestLogHandler::new().exists_log_containing(&format!(
+            "WARN: {test_name}: Total of {} wei in MASQ was ordered while the consuming wallet held \
+            only {}", service_fee_totally_required_minor.separate_with_commas(),
+            (cw_service_fee_balance_minor - 2).separate_with_commas()
+        ));
+    }
+
+    #[test]
+    fn construction_of_error_context_with_accounts_dumped_works() {
+        let balance_1 = 1234567;
+        let mut account_1 = make_weighed_account(123);
+        account_1
+            .analyzed_account
+            .qualified_as
+            .bare_account
+            .balance_wei = balance_1;
+        let balance_2 = 999888777;
+        let mut account_2 = make_weighed_account(345);
+        account_2
+            .analyzed_account
+            .qualified_as
+            .bare_account
+            .balance_wei = balance_2;
+        let weighted_accounts = vec![account_1, account_2];
+
+        let result = LateServiceFeeSingleTxErrorFactory::new(&weighted_accounts);
+
+        assert_eq!(
+            result,
+            LateServiceFeeSingleTxErrorFactory {
+                original_number_of_accounts: 2,
+                original_service_fee_required_total_minor: balance_1 + balance_2
+            }
+        )
+    }
+
+    fn double_mock_results_queue(mock: DisqualificationGaugeMock) -> DisqualificationGaugeMock {
+        let originally_prepared_results = (0..2)
```

Comment:
Let's call the variable `popped_results`


`2` -> `len_of_queue`
`4` -> `2 * len_of_queue`